src/crimson/common/tri_mutex.cc

```diff
@@ -181,12 +183,22 @@ bool tri_mutex::is_acquired() const
   }
 }
 
-void tri_mutex::wake()
+void tri_mutex::wake(type_t type)
 {
   LOG_PREFIX(tri_mutex::wake());
   DEBUGDPP("", *this);
-  assert(!readers && !writers && !exclusively_used);
-  type_t type = type_t::none;
+#ifndef NDEBUG
+  switch(type) {
+  case type_t::none:
+    assert(!readers && !writers && !exclusively_used);
+  case type_t::read:
+    assert(!writers && !exclusively_used);
+  case type_t::write:
+    assert(!readers && !exclusively_used);
+  default:
+    ceph_abort();
+  }
+#endif
```

Comment:
I cannot follow the commit message, looks it is simply adding some asserts, how does it fix an issue?

And do you think it is better to implement asserts in each unlock/demote methods instead?

Reply:
> I cannot follow the commit message, looks it is simply adding some asserts, how does it fix an issue?
> 
> And do you think it is better to implement asserts in each unlock/demote methods instead?

Sorry, I'll improve the commit message which is indeed misleading at present. The issue was actually that when demoting excl locks tri_mutex doesn't wake waiters, so a deadlock is possible in the following case:

1. A client request `REQ`  tries to acquire the `READ` lock on the head obc and finds that the head obc is missing in memory, so it acquires the `EXCL` lock on the head obc and loads the metadata from disk;
2. A clone object `O` is added to `PGRecovery::recovering`; `PGRecovery` tries to acquire the `READ` lock  on the head obc, which makes it wait on the previous `EXCL` lock;
3. `REQ` finishes loading the metadata, demotes the lock to `READ`, tries to recover the clone object `O` and finds that it's already being recovered, so it waits on it;

Since the `demote` process doesn't wake up waiters, this is a deadlock.

Reply:
This PR make the `demote` wake up waiters.

Reply:
Ah, just realized that `wake()` is added to the demote methods.

Reply:
Nit, probably it is more straightforward to move the asserts to each unlock/demote methods.

Reply:
The "type" param here is kind of confusing.  As used, it appears that you pass a type corresponding to what you expect to find.  That doesn't really seem like a property of wake -- rather it seems like a detail of the line(s) immediately prior to wake().  I think I'd rather these asserts just be placed before the call to wake(), even if the type_t::none one ends up duplicated.

Reply:
Looks reasonable to me otherwise -- demote_to_* should wake().

Reply:
All done, please take a look again.